python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,458 @@
+                }
+    doc = {"summary": "n add function",
+           "description": "add N number of arrays"
+           }
+    pc.register_scalar_function(n_add, "n_add", doc,
+                                in_types, pa.int64())
+
+    func = pc.get_function("n_add")
+
+    assert func.name == "n_add"
+    error_msg = "VarArgs function 'n_add' needs at least 2 arguments"
+    with pytest.raises(pa.lib.ArrowInvalid, match=error_msg):
+        pc.call_function("n_add", [pa.array([1, 10]),
+                                   ])
+
+
+def test_function_doc_validation():
+
+    def unary_scalar_function(ctx, scalar):
+        return pc.call_function("add", [scalar, 1])
+
+    # validate arity
+    in_types = {"scalar": pc.InputType.scalar(pa.int64())}
+    out_type = pa.int64()
+
+    # doc with no summary
+    func_doc = {
+        "description": "desc"
+    }
+
+    expected_expr = "Function doc must contain a summary"
+
+    with pytest.raises(ValueError, match=expected_expr):
+        pc.register_scalar_function(unary_scalar_function, "no_summary",
+                                    func_doc, in_types,
+                                    out_type)
+
+    # doc with no decription
+    func_doc = {
+        "summary": "test summary"
+    }
+
+    expected_expr = "Function doc must contain a description"
+
+    with pytest.raises(ValueError, match=expected_expr):
+        pc.register_scalar_function(unary_scalar_function, "no_desc",
+                                    func_doc, in_types,
+                                    out_type)
+
+    # doc with empty dictionary
+    func_doc = {}
+    expected_expr = "Function doc must contain a summary"
+    with pytest.raises(ValueError, match=expected_expr):
+        pc.register_scalar_function(unary_scalar_function, "empty_dictionary",
+                                    func_doc, in_types,
+                                    out_type)
+
+
+def test_non_uniform_input_udfs():
+
+    def unary_scalar_function(ctx, scalar1, array1, scalar2):
+        coeff = pc.call_function("add", [scalar1, scalar2])
+        return pc.call_function("multiply", [coeff, array1])
+
+    in_types = {"scalar1": pc.InputType.scalar(pa.int64()),
+                "scalar2": pc.InputType.array(pa.int64()),
+                "scalar3": pc.InputType.scalar(pa.int64()),
+                }
+    func_doc = {
+        "summary": "multi type udf",
+        "description": "desc"
+    }
+    pc.register_scalar_function(unary_scalar_function,
+                                "multi_type_udf", func_doc,
+                                in_types,
+                                pa.int64())
+
+    res = pc.call_function("multi_type_udf",
+                           [pa.scalar(10), pa.array([1, 2, 3]), pa.scalar(20)])
+    assert res == pa.array([30, 60, 90])
+
+
+def test_nullary_functions():
+
+    def gen_random(ctx):
+        import random
+        val = random.randint(0, 10)
+        return pa.scalar(val)
```

Comment:
Could you keep a single convention of putting all these functions at the toplevel? Currently, some are at the toplevel and some are local inside test functions...

Reply:
Nice idea. we can remove if there are any redundant functions.